File: src/org/kde/kdeconnect/Plugins/SystemVolumePlugin/SystemVolumeFragment.java
/*
 * SPDX-FileCopyrightText: 2018 Nicolas Fella <nicolas.fella@gmx.de>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
 */

package org.kde.kdeconnect.Plugins.SystemVolumePlugin;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Consumer;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import org.kde.kdeconnect.Helpers.VolumeHelperKt;
import org.kde.kdeconnect.KdeConnect;
import org.kde.kdeconnect.Plugins.MprisPlugin.MprisPlugin;
import org.kde.kdeconnect.Plugins.MprisPlugin.VolumeKeyListener;
import org.kde.kdeconnect_tp.R;
import org.kde.kdeconnect_tp.databinding.ListItemSystemvolumeBinding;
import org.kde.kdeconnect_tp.databinding.SystemVolumeFragmentBinding;

import java.util.ArrayList;
import java.util.List;

public class SystemVolumeFragment
        extends Fragment
        implements Sink.UpdateListener, SystemVolumePlugin.SinkListener, VolumeKeyListener {

    private SystemVolumePlugin plugin;
    private RecyclerSinkAdapter recyclerAdapter;
    private boolean tracking;
    private final Consumer<Boolean> trackingConsumer = aBoolean -> tracking = aBoolean;
    private SystemVolumeFragmentBinding systemVolumeFragmentBinding;

    public static SystemVolumeFragment newInstance(String deviceId) {
        SystemVolumeFragment systemvolumeFragment = new SystemVolumeFragment();

        Bundle arguments = new Bundle();
        arguments.putString(MprisPlugin.DEVICE_ID_KEY, deviceId);

        systemvolumeFragment.setArguments(arguments);

        return systemvolumeFragment;
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState
    ) {

        if (systemVolumeFragmentBinding == null) {
            systemVolumeFragmentBinding = SystemVolumeFragmentBinding.inflate(inflater);

            RecyclerView recyclerView = systemVolumeFragmentBinding.audioDevicesRecycler;

            int gap = requireContext().getResources().getDimensionPixelSize(R.dimen.activity_vertical_margin);
            recyclerView.addItemDecoration(new ItemGapDecoration(gap));
            recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));

            recyclerAdapter = new RecyclerSinkAdapter();
            recyclerView.setAdapter(recyclerAdapter);
        }

        connectToPlugin(getDeviceId());

        return systemVolumeFragmentBinding.getRoot();
    }

    @Override
    public void onDestroyView() {
        disconnectFromPlugin(getDeviceId());
        super.onDestroyView();
    }

    @Override
    public void updateSink(@NonNull final Sink sink) {

        // Don't set progress while the slider is moved
        if (!tracking) {

            requireActivity().runOnUiThread(() -> recyclerAdapter.notifyDataSetChanged());
        }
    }

    private void connectToPlugin(final String deviceId) {
        SystemVolumePlugin plugin = KdeConnect.getInstance().getDevicePlugin(deviceId, SystemVolumePlugin.class);
        if (plugin == null) {
            return;
        }
        this.plugin = plugin;
        plugin.addSinkListener(SystemVolumeFragment.this);
        plugin.requestSinkList();
    }

    private void disconnectFromPlugin(final String deviceId) {
        SystemVolumePlugin plugin = KdeConnect.getInstance().getDevicePlugin(deviceId, SystemVolumePlugin.class);
        if (plugin == null) {
            return;
        }
        plugin.removeSinkListener(SystemVolumeFragment.this);
    }

    @Override
    public void sinksChanged() {

        for (Sink sink : plugin.getSinks()) {
            sink.addListener(SystemVolumeFragment.this);
        }

        requireActivity().runOnUiThread(() -> {
            List<Sink> newSinks = new ArrayList<>(plugin.getSinks());
            recyclerAdapter.submitList(newSinks);
        });
    }

    @Override
    public void onVolumeUp() {
        updateDefaultSinkVolume(5);
    }

    @Override
    public void onVolumeDown() {
        updateDefaultSinkVolume(-5);
    }

    private void updateDefaultSinkVolume(int percent) {
        if (plugin == null) return;

        Sink defaultSink = SystemVolumeUtilsKt.getDefaultSink(plugin);
        if (defaultSink == null) return;

        int newVolume = VolumeHelperKt.calculateNewVolume(
                defaultSink.getVolume(),
                defaultSink.getMaxVolume(),
                percent
        );

        if (defaultSink.getVolume() == newVolume) return;

        plugin.sendVolume(defaultSink.getName(), newVolume);
    }

    private String getDeviceId() {
        return requireArguments().getString(MprisPlugin.DEVICE_ID_KEY);
    }

    private class RecyclerSinkAdapter extends ListAdapter<Sink, SinkItemHolder> {

        public RecyclerSinkAdapter() {
            super(new SinkItemCallback());
        }

        @NonNull
        @Override
        public SinkItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

            LayoutInflater inflater = getLayoutInflater();
            ListItemSystemvolumeBinding viewBinding = ListItemSystemvolumeBinding.inflate(inflater, parent, false);

            return new SinkItemHolder(viewBinding, plugin, trackingConsumer);
        }

        @Override
        public void onBindViewHolder(@NonNull SinkItemHolder holder, int position) {
            holder.bind(getItem(position));
        }
    }
}

File: src/org/kde/kdeconnect/Plugins/SystemVolumePlugin/SystemVolumeFragment.kt
/*
 * SPDX-FileCopyrightText: 2018 Nicolas Fella <nicolas.fella@gmx.de>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
 */
package org.kde.kdeconnect.Plugins.SystemVolumePlugin

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.util.Consumer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ListAdapter
import org.kde.kdeconnect.Helpers.calculateNewVolume
import org.kde.kdeconnect.KdeConnect
import org.kde.kdeconnect.Plugins.MprisPlugin.MprisPlugin
import org.kde.kdeconnect.Plugins.MprisPlugin.VolumeKeyListener
import org.kde.kdeconnect.Plugins.SystemVolumePlugin.SystemVolumePlugin.SinkListener
import org.kde.kdeconnect.base.BaseFragment
import org.kde.kdeconnect.extensions.setupBottomPadding
import org.kde.kdeconnect_tp.R
import org.kde.kdeconnect_tp.databinding.ListItemSystemvolumeBinding
import org.kde.kdeconnect_tp.databinding.SystemVolumeFragmentBinding

class SystemVolumeFragment : BaseFragment<SystemVolumeFragmentBinding>(),
    Sink.UpdateListener, SinkListener, VolumeKeyListener {

    private lateinit var plugin: SystemVolumePlugin
    private lateinit var recyclerAdapter: RecyclerSinkAdapter
    private var tracking = false
    private val trackingConsumer = Consumer { aBoolean: Boolean -> tracking = aBoolean }

    override fun getActionBarTitle() = getString(R.string.open_mpris_controls)

    override fun onInflateBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): SystemVolumeFragmentBinding {
        return SystemVolumeFragmentBinding.inflate(inflater, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recyclerAdapter = RecyclerSinkAdapter()
        binding.audioDevicesRecycler.apply {
            layoutManager = LinearLayoutManager(requireContext())
            addItemDecoration(ItemGapDecoration(resources.getDimensionPixelSize(R.dimen.activity_vertical_margin)))
            adapter = recyclerAdapter
            setupBottomPadding()
        }
        connectToPlugin(deviceId)
    }

    override fun onDestroyView() {
        disconnectFromPlugin(deviceId)
        super.onDestroyView()
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun updateSink(sink: Sink) {
        // Don't set progress while the slider is moved
        if (!tracking) {
            requireActivity().runOnUiThread {
                if (::recyclerAdapter.isInitialized) {
                    recyclerAdapter.notifyDataSetChanged()
                }
            }
        }
    }

    private fun connectToPlugin(deviceId: String?) {
        val plugin = KdeConnect.getInstance().getDevicePlugin(
            deviceId,
            SystemVolumePlugin::class.java
        )
        if (plugin == null) {
            return
        }
        this.plugin = plugin
        plugin.addSinkListener(this@SystemVolumeFragment)
        plugin.requestSinkList()
    }

    private fun disconnectFromPlugin(deviceId: String?) {
        val plugin = KdeConnect.getInstance().getDevicePlugin(
            deviceId,
            SystemVolumePlugin::class.java
        )
        if (plugin == null) {
            return
        }
        plugin.removeSinkListener(this@SystemVolumeFragment)
    }

    override fun sinksChanged() {
        if (!::plugin.isInitialized || !::recyclerAdapter.isInitialized) {
            return
        }
        for (sink in plugin.sinks) {
            sink.addListener(this@SystemVolumeFragment)
        }
        requireActivity().runOnUiThread {
            val newSinks: List<Sink> = ArrayList(plugin.sinks)
            recyclerAdapter.submitList(newSinks)
        }
    }

    override fun onVolumeUp() {
        updateDefaultSinkVolume(5)
    }

    override fun onVolumeDown() {
        updateDefaultSinkVolume(-5)
    }

    private fun updateDefaultSinkVolume(percent: Int) {
        if (!::plugin.isInitialized) {
            return
        }

        val defaultSink = getDefaultSink(plugin) ?: return

        val newVolume = calculateNewVolume(
            defaultSink.volume,
            defaultSink.maxVolume,
            percent
        )

        if (defaultSink.volume == newVolume) return

        plugin.sendVolume(defaultSink.name, newVolume)
    }

    private val deviceId: String?
        get() = arguments?.getString(MprisPlugin.DEVICE_ID_KEY)

    private inner class RecyclerSinkAdapter : ListAdapter<Sink?, SinkItemHolder>(SinkItemCallback()) {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SinkItemHolder {
            val viewBinding = ListItemSystemvolumeBinding.inflate(layoutInflater, parent, false)
            return SinkItemHolder(viewBinding, plugin, trackingConsumer)
        }

        override fun onBindViewHolder(holder: SinkItemHolder, position: Int) {
            holder.bind(getItem(position))
        }
    }

    companion object {
        fun newInstance(deviceId: String?): SystemVolumeFragment {
            val systemVolumeFragment = SystemVolumeFragment()

            val arguments = Bundle()
            arguments.putString(MprisPlugin.DEVICE_ID_KEY, deviceId)

            systemVolumeFragment.arguments = arguments

            return systemVolumeFragment
        }
    }
}
